pkg/cloud/azureprovider.go

```diff
@@ -740,10 +839,55 @@ func (az *Azure) AllNodePricing() (interface{}, error) {
 func (az *Azure) NodePricing(key Key) (*Node, error) {
 	az.DownloadPricingDataLock.RLock()
 	defer az.DownloadPricingDataLock.RUnlock()
+
 	azKey, ok := key.(*azureKey)
 	if !ok {
 		return nil, fmt.Errorf("azure: NodePricing: key is of type %T", key)
 	}
+	config, _ := az.GetConfig()
+	if config.SpotLabel != "" && config.SpotLabelValue != "" {
+		labelsMapStr := fmt.Sprintf("%s", azKey)
+		spotLabelMapStr := fmt.Sprintf(" %s:%s ", config.SpotLabel, config.SpotLabelValue)
+		if strings.Contains(labelsMapStr, spotLabelMapStr) {
+			features := strings.Split(azKey.Features(), ",")
+			spotFeatures := strings.ReplaceAll(azKey.Features(), ",ondemand", ",spot")
+			if n, ok := az.Pricing[spotFeatures]; ok {
+				klog.V(4).Infof("Returning pricing for node %s: %+v from key %s", azKey, n, spotFeatures)
+				if azKey.isValidGPUNode() {
+					n.Node.GPU = "1" // TODO: support multiple GPUs
+				}
+				return n.Node, nil
+			}
+			klog.V(1).Infof("[Info] found spot instance, trying to get retail price for %s: %s, ", spotFeatures, azKey)
+
+			spotCost, err := getRetailPrice(features[0], features[1], config.CurrencyCode, true)
+			if err != nil {
+				log.DedupedWarningf(5, "failed to retrieve spot retail pricing")
+			} else {
+				gpu := ""
+				allPrices, err := az.AllNodePricing()
+				if err == nil {
+					if azKey.isValidGPUNode() {
+						gpu = "1"
+					}
+					allPrices.(map[string]*AzurePricing)[spotFeatures] = &AzurePricing{
+						Node: &Node{
+							Cost:      spotCost,
+							UsageType: "spot",
+							GPU:       gpu,
+						},
+					}
+					az.Pricing = allPrices.(map[string]*AzurePricing)
```

Comment:
Is there any mutex or concurrency control on `az.Pricing` that should be checked here? I see a mutex being used in `AllNodePricing` for something like this.

--

Also, what if there are 2 different unknown node types? Won't this set of logic do the following?
1. create a new allprices with node type 1
2. update `az.Pricing` with the new allprices with node type 1
3. lookup node type 2 and get back in this logic
4. create a new allprices with node type 2 (and not node type 1)
5. update `az.Pricing` with the new allprices that has node type 1 but not node type 2

Would it be better to just merge the current `az.Pricing` map with the new, updated map?

Reply:
The same mutex is being locked for this function
```
func (az *Azure) NodePricing(key Key) (*Node, error) {
	az.DownloadPricingDataLock.RLock()
	defer az.DownloadPricingDataLock.RUnlock()
```

What I am unsure of is since `az.AllNodePricing()` is called here and contains a `defer az.DownloadPricingDataLock.RUnlock()` it seems like it is unlocking the mutex prematurely

Reply:
That's only a read lock. I think you need to be using `.Lock()` when editing `az.Pricing` and `.RLock()` when reading.